lib/spack/spack/installer.py

```diff
@@ -400,7 +400,11 @@ def dump_packages(spec, path):
 
             # There's no provenance installed for the source package.  Skip it.
             # User can always get something current from the builtin repo.
-            if not os.path.isdir(source_repo_root):
+            # If it's external, treat it as current because we cannot guarantee
```

Comment:
What does "treat it as current" mean?

I think the change is fine but I'm confused where this code fails for external packages installed by Spack (even if the instance doing the installation doesn't have access to the repo - in fact I thought the point of this was to extract the standalone repo constructed in the install directory). In any case, if you point out which line raises an error that might clear things up for me.

Reply:
I've pinged you in the conversation on slack, maybe once you read that we can iterate on a better comment so we don't confuse other maintainers?

Reply:
Based on that Slack conversation, it is the last line of this loop body that is failing:

```
fs.install_tree(source_pkg_dir, dest_pkg_dir)
```

because `source_pkg_dir` doesn't exist. This appears to be because

* Spec.concretize automatically assigns namespaces based on searching for
  packages in the repositories that are available to it (e.g. rather than
  inspecting the installation directory for the standalone repo)
* External packages are not assigned a namespace, even if they were built
  by Spack.

Put together, this means that external Spack-built packages are likely to
get the wrong namespace. In fact, if the package doesn't exist in the repo of
the local Spack instance, I think the following line would fail in
Spack.concretize:

```
if s.namespace is None:
    s.namespace = spack.repo.path.repo_for_pkg(s.name).namespace
```

i.e. this only got as far as it did because `CMake` existed in the local
repository, and their would have been an earlier failure for any package
where the package name did not exist in the local repository.

So this may also be possible to resolve by assigning the namespace properly.
We could likely avoid additional problems by leaving `namespace=None` for
external packages. IMO if it's not cleaned up here, it's worth mentioning in
the comment that external packages do not have properly-assigned namespaces
and that therefore we cannot make use of their repository (even if it is
well-formed).

Reply:
So I would revise the comment in two ways:

* Replace "treat it as current" with "skip it"
* we do know the repo, but we aren't accessing it properly right now and skipping it is a quick fix

e.g.

> Skip adding external packages to repository since their namespace may not be properly assigned

Admittedly, it is also possible that the directory layout could differ for sufficiently divergent Spack versions and that may be a reason to skip externals (this is unlikely though since most of the details of the prefix would already be recorded in the external entry).